### ISSUE TITLE
SDPA returns 0 output and 0 gradient with corner case input

### ISSUE THREAD

**Issue by kaixuanliu** [intel/torch-xpu-ops#3710]
### 🐛 Describe the bug

We meet a corner case produced by the tiny Qwen2.5-VL vision attention after RoPE: hidden_size=16, num_heads=16 -> head_dim=1, while RoPE cos/sin have last dimension 0. As a result q/k become (..., 0), but v remains (..., 1). With this input SDPA on XPU will return 0 output and 0 gradient, while on CUDA it is OK and w/ eager    mode it is OK as well.
```
import torch
import torch.nn.functional as F


def eager_attention(query, key, value, scale):
    scores = torch.matmul(query, key.transpose(-2, -1)) * scale
    probs = torch.softmax(scores, dim=-1, dtype=torch.float32).to(query.dtype)
    return torch.matmul(probs, value)


def run(use_sdpa, device):
    torch.manual_seed(0)
    query = torch.empty((1, 16, 16, 0), device=device, requires_grad=True)
    key = torch.empty((1, 16, 16, 0), device=device, requires_grad=True)
    value = torch.randn((1, 16, 16, 1), device=device, requires_grad=True)
    grad_output = torch.randn((1, 16, 16, 1), device=device)

    if use_sdpa:
        output = F.scaled_dot_product_attention(
            query,
            key,
            value,
            dropout_p=0.0,
            is_causal=False,
            scale=1.0,
        )
    else:
        output = eager_attention(query, key, value, scale=1.0)

    output.backward(grad_output)
    return output.detach().abs().sum().item(), value.grad.detach().abs().sum().item()


def compare(device):
    sdpa_output_sum, sdpa_value_grad_sum = run(use_sdpa=True, device=device)
    eager_output_sum, eager_value_grad_sum = run(use_sdpa=False, device=device)

    print(f"\n{device}")
    print(f"sdpa : output_abs_sum={sdpa_output_sum:.9e}, value_grad_abs_sum={sdpa_value_grad_sum:.9e}")
    print(f"eager: output_abs_sum={eager_output_sum:.9e}, value_grad_abs_sum={eager_value_grad_sum:.9e}")

    if sdpa_value_grad_sum == 0.0 and eager_value_grad_sum > 0.0:
        print("BUG: SDPA returns zero value gradient while eager attention returns a non-zero gradient.")


def main():
    print(f"torch: {torch.__version__}")
    if hasattr(torch, "xpu") and torch.xpu.is_available():
        print(f"xpu device: {torch.xpu.get_device_name(0)}")
        compare("xpu")


if __name__ == "__main__":
    main()
```

### Versions

```
PyTorch version: 2.12.0+xpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.3 (main, Mar 23 2026, 19:04:32) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.14.0-35-generic-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: True
XPU used to build PyTorch: 20250302
Intel GPU driver version:
* libze1:       1.28.2-1~24.04~ppa1
* intel-opencl-icd:     26.14.37833.4-1~24.04~ppa1
Intel GPU models onboard:
N/A
Intel GPU models detected:
* [0] _XpuDeviceProperties(name='Intel(R) Arc(TM) Pro B60 Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE211, uuid=868011e2-0000-0000-2900-000000000000, driver_version='1.15.37833+4', total_memory=24480MB, local_mem_size=128KB, max_compute_units=160, memory_clock_rate=2400MHz, memory_bus_width=64-bit, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [1] _XpuDeviceProperties(name='Intel(R) Arc(TM) Pro B60 Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE211, uuid=868011e2-0000-0000-8300-000000000000, driver_version='1.15.37833+4', total_memory=24480MB, local_mem_size=128KB, max_compute_units=160, memory_clock_rate=2400MHz, memory_bus_width=64-bit, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
* [2] _XpuDeviceProperties(name='Intel(R) Arc(TM) Pro B60 Graphics', platform_name='Intel(R) oneAPI Unified Runtime over Level-Zero V2', type='gpu', device_id=0xE211, uuid=868011e2-0000-0000-aa00-000000000000, driver_version='1.15.37833+4', total_memory=24480MB, local_mem_size=128KB, max_compute_units=160, memory_clock_rate=2400MHz, memory_bus_width=64-bit, gpu_eu_count=160, gpu_subslice_count=20, max_work_group_size=1024, max_num_sub_groups=64, sub_group_sizes=[16 32], has_fp16=1, has_fp64=1, has_atomic64=1)
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           52 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  224
On-line CPU(s) list:                     0-223
Vendor ID:                               GenuineIntel
BIOS Vendor ID:                          Intel(R) Corporation
Model name:                              Intel(R) Xeon(R) Platinum 8480+
BIOS Model name:                         Intel(R) Xeon(R) Platinum 8480+  CPU @ 2.0GHz
BIOS CPU family:                         179
CPU family:                              6
Model:                                   143
Thread(s) per core:                      2
Core(s) per socket:                      56
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      22%
CPU max MHz:                             3800.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4000.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 intel_ppin cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect user_shstk avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               5.3 MiB (112 instances)
L1i cache:                               3.5 MiB (112 instances)
L2 cache:                                224 MiB (112 instances)
L3 cache:                                210 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-55,112-167
NUMA node1 CPU(s):                       56-111,168-223
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] apollo-torch==1.0.3
[pip3] dpcpp-cpp-rt==2025.3.2
[pip3] galore-torch==1.0
[pip3] impi-rt==2021.17.2
[pip3] intel-cmplr-lib-rt==2025.3.2
[pip3] intel-cmplr-lib-ur==2025.3.2
[pip3] intel-cmplr-lic-rt==2025.3.2
[pip3] intel-opencl-rt==2025.3.2
[pip3] intel-openmp==2025.3.2
[pip3] intel-pti==0.16.0
[pip3] intel-sycl-rt==2025.3.2
[pip3] mkl==2025.3.1
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.2.6
[pip3] oneccl==2021.17.2
[pip3] oneccl-devel==2021.17.2
[pip3] onemkl-license==2025.3.1
[pip3] onemkl-sycl-blas==2025.3.1
[pip3] onemkl-sycl-dft==2025.3.1
[pip3] onemkl-sycl-lapack==2025.3.1
[pip3] onemkl-sycl-rng==2025.3.1
[pip3] onemkl-sycl-sparse==2025.3.1
[pip3] onnx==1.21.0
[pip3] pytorch-msssim==1.0.0
[pip3] tbb==2022.3.1
[pip3] tcmlib==1.4.1
[pip3] torch==2.12.0+xpu
[pip3] torchao==0.17.0
[pip3] torchaudio==2.11.0+xpu
[pip3] torchcodec==0.12.0+cpu
[pip3] torchdata==0.11.0
[pip3] torchvision==0.27.0+xpu
[pip3] triton==3.7.0
[pip3] triton-xpu==3.7.1
[pip3] umf==1.0.3
```

**Comment by kaixuanliu** [comment]
After I upgrade torch to 2.12 for cuda as well, it has the same behavior. Close this issue.
